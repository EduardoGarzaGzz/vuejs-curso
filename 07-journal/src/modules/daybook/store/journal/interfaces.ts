export interface IJournalState {
    isLoading?: boolean,
    entries?: any[]
}
