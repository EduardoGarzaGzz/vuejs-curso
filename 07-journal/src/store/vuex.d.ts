import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Store }         from 'vuex';

declare module '@vue/runtime-core' {

    interface State {
        journal: IJournalState;
    }

    interface ComponentCustomProperties {
        $store: Store<State>;
    }

}
