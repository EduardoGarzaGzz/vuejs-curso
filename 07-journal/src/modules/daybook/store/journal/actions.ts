import journalApi        from '@/api/journalApi';
import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Action }        from 'vuex';

export const loadEntries: Action<IJournalState, IJournalState> = async ( { commit } ) => {
    const { data }     = await journalApi.get( `/entries.json` );
    const entries: any = [];

    if ( !data ) {
        commit( 'setEntries', [] );
        return;
    }

    for ( const id of Object.keys( data ) ) {
        entries.push( {
            id,
            ...data[ id ]
        } );
    }

    commit( 'setEntries', entries );
};

export const updateEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    const { id, date, picture, text } = entry;
    const { data }                    = await journalApi.put( `/entries/${ id }.json`, {
        date,
        picture,
        text
    } );
    commit( 'updateEntry', {
        id,
        ...data
    } );
};

export const createEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    const { date, picture, text } = entry;
    const dataToSave              = { date, picture, text };
    const { data }                = await journalApi.post( `/entries.json`, dataToSave );
    commit( 'addEntry', {
        id: data.name,
        ...dataToSave
    } );
    return data.name;
};

export const deleteEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    await journalApi.delete( `/entries/${ entry.id }.json` );
    commit( 'deleteEntry', entry );
};
