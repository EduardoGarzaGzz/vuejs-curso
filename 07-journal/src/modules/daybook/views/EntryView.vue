<template>
	<template v-if="entry">
		<div class="entry-title d-flex justify-content-between p-2">
			<div>
				<span class="text-success fs-5 fw-bold">{{ day }}</span>
				<span class="mx-1 fs-5">{{ month }}</span>
				<span class="mx-2 fw-light">{{ yearDay }}</span>
			</div>
			<div>
				<button v-if="entry.id"
								class="btn btn-danger mx-2"
								@click="onDeleteEntry">Borrar <em class="fa fa-trash-alt"></em></button>
				<button class="btn btn-primary mx-2">Subir foto <em class="fa fa-upload"></em></button>
			</div>
		</div>
		<hr>
		<div class="d-flex flex-column px-3 h-75">
			<textarea v-model="entry.text" placeholder="Que sucediÃ³ hoy?"></textarea>
		</div>
		<img alt="entry-picture" class="img-thumbnail" src="http://qnimate.com/wp-content/uploads/2014/03/images2.jpg">
	</template>
	<Fab icon="fa-save" @on:click="saveEntry" />
</template>

<script lang="js">
import { defineAsyncComponent }   from 'vue';
import { mapActions, mapGetters } from "vuex";
import getDayMonthYear            from "@/modules/daybook/helpers/getDayMonthYear.ts";

export default {
	name      : 'EntryView',
	props     : {
		id: {
			type    : String,
			required: true
		}
	},
	components: {
		Fab: defineAsyncComponent( () => import('@/modules/daybook/components/Fab.vue') )
	},
	data() {
		return {
			entry: {
				text: ''
			}
		}
	},
	computed: {
		...mapGetters( 'journal', {
			getEntriesById: 'getEntriesById'
		} ),
		day() {
			const { day } = getDayMonthYear( this.entry.date );
			return day;
		},
		month() {
			const { month } = getDayMonthYear( this.entry.date );
			return month;
		},
		yearDay() {
			const { yearDay } = getDayMonthYear( this.entry.date );
			return yearDay;
		}
	},
	methods : {
		...mapActions( 'journal', {
			updateEntry: "updateEntry",
			createEntry: "createEntry",
			deleteEntry: "deleteEntry"
		} ),
		loadEntry() {
			let entry = null;
			if ( this.id === 'new' ) {
				this.entry = {
					text: '',
					date: new Date().getTime()
				};
			} else {
				entry = this.getEntriesById( this.id );
				
				if ( !entry ) {
					this.$router.push( { name: 'no-entry' } );
					return;
				}
				
				this.entry = entry;
			}
		},
		async saveEntry() {
			if ( this.entry.id ) {
				await this.updateEntry( this.entry );
			} else {
				const id = await this.createEntry( this.entry );
				await this.$router.push( { name: 'entry', params: { id } } );
			}
		},
		async onDeleteEntry() {
			await this.deleteEntry( this.entry );
			await this.$router.push( { name: 'no-entry' } );
		}
	},
	created() {
		this.loadEntry();
	},
	watch: {
		id() {
			this.loadEntry();
		}
	}
};
</script>

<style lang="scss" scoped>

textarea {
	font-size: 20px;
	border: none;
	height: 100%;
	
	&:focus {
		outline: none;
	}
}

img {
	width: 200px;
	position: fixed;
	bottom: 150px;
	right: 20px;
	box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}

</style>
