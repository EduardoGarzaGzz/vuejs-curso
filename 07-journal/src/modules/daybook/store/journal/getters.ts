import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Getter }        from 'vuex';

export const getEntriesByTerm: Getter<IJournalState, IJournalState> = ( state ) => ( term: string | null = null ): any => {
    const list: any = state.entries;

    if ( !term ) return list;
    return list.filter( ( entry: any ) => entry.text.toLowerCase().includes( term ) );
};


export const getEntriesById: Getter<IJournalState, IJournalState> = ( state ) => ( id = '' ) => {
    const entry = ( state.entries as any ).find( ( entry: any ) => entry.id === id );

    if ( !entry ) return undefined;

    return { ...entry };
};
