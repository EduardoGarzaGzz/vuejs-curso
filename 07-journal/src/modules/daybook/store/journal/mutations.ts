import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Mutation }      from 'vuex';

export const setEntries: Mutation<IJournalState> = ( state, entries: any ) => {
    state.entries   = [ ...( state.entries as any ), ...entries ];
    state.isLoading = false;
};

export const updateEntry: Mutation<IJournalState> = ( state, entryUpdated: any ) => {
    const idx = ( state.entries as any ).map( ( e: any ) => e.id ).indexOf( entryUpdated.id );

    if ( idx !== -1 ) {
        ( state.entries as any )[ idx ] = entryUpdated;
    } else {
        ( state.entries as any ).push( { ...entryUpdated } );
    }
};

export const addEntry: Mutation<IJournalState> = ( state, entry ) => {
    state.entries = [ entry, ...state.entries ];
};

export const deleteEntry: Mutation<IJournalState> = ( state, entry ) => {
    state.entries = state.entries.filter( e => e.id !== entry.id );
};
