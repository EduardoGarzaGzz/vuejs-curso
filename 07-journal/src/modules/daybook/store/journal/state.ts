import { IJournalState } from '@/modules/daybook/store/journal/interfaces';


export const journalState = (): IJournalState => ( {
    isLoading: true,
    entries  : [
        {
            id     : new Date().getTime(),
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 1,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 2,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 3,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        }
    ]
} );
