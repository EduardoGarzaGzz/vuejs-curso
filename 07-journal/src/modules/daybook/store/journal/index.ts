import * as actions              from '@/modules/daybook/store/journal/actions';
import * as getters              from '@/modules/daybook/store/journal/getters';
import { IJournalState }         from '@/modules/daybook/store/journal/interfaces';
import * as mutations            from '@/modules/daybook/store/journal/mutations';
import { journalState as state } from '@/modules/daybook/store/journal/state';
import { Module }                from 'vuex';

export const journalModule: Module<IJournalState, IJournalState> = {
    namespaced: true,
    state,
    actions,
    mutations,
    getters
};
