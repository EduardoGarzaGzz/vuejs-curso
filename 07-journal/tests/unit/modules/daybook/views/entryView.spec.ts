import { journalModule } from '@/modules/daybook/store/journal';
import EntryView         from '@/modules/daybook/views/EntryView.vue';
import { shallowMount }  from '@vue/test-utils';
import Swal              from 'sweetalert2';
import { createStore }   from 'vuex';
import { journalState }  from '../../../mock-data/test-journal-state';

jest.mock( 'sweetalert2', () => ( {
	fire       : jest.fn(),
	showLoading: jest.fn(),
	close      : jest.fn()
} ) );

const createVuexStore = ( initialStore: any ) => createStore( {
	modules: {
		journal: {
			...journalModule,
			state: { ...initialStore }
		}
	}
} );

describe( 'Test View: EntryView.vue', function() {
	const store    = createVuexStore( journalState );
	store.dispatch = jest.fn();

	const mockRouter = {
		push: jest.fn()
	};

	let wrapper: any;

	beforeEach( () => {
		jest.clearAllMocks();

		wrapper = shallowMount( EntryView, {
			props : {
				id: '-Msr8-0x5wA7UhFR7dIB'
			},
			global: {
				mocks  : {
					$router: mockRouter
				},
				plugins: [ store ]
			}
		} );
	} );

	it( 'debe de sacar al usuario porque el id no existe', () => {
		const wrapper = shallowMount( EntryView, {
			props : {
				id: 'Este ID no existe'
			},
			global: {
				mocks  : {
					$router: mockRouter
				},
				plugins: [ store ]
			}
		} );

		expect( mockRouter.push ).toHaveBeenCalledWith( { name: 'no-entry' } );
	} );

	it( 'debe de mostrar la entrada correcta', () => {
		expect( wrapper.html() ).toMatchSnapshot();
		expect( mockRouter.push ).not.toHaveBeenCalledWith( { name: 'no-entry' } );
	} );

	it( 'debe de borrar la entrada y salir', async ( done ) => {
		//@ts-ignore
		Swal.fire.mockReturnValueOnce( Promise.resolve( { isConfirmed: true } ) );

		await wrapper.find( '.btn-danger' ).trigger( 'click' );
		expect( Swal.fire ).toHaveBeenCalledWith( {
			title            : 'Esta seguro?',
			text             : 'Una vez borrado, no se puede recuperar',
			showDenyButton   : true,
			confirmButtonText: 'Si, estoy seguro'
		} );

		setTimeout( () => {
			expect( store.dispatch ).toHaveBeenCalledWith( 'journal/deleteEntry', '-Msr8-0x5wA7UhFR7dIB' );
			expect( mockRouter.push ).toHaveBeenCalledWith( { name: 'no-entry' } );
			done();
		}, 1 );
	} );
} );
