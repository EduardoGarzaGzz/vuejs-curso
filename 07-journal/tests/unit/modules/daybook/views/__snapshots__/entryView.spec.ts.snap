// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test View: EntryView.vue debe de mostrar la entrada correcta 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-5 fw-bold">23</span><span class="mx-1 fs-5">Julio</span><span class="mx-2 fw-light">2021, Viernes</span></div>
  <div><input accept="image/png, image/jpeg" type="file" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <em class="fa fa-trash-alt"></em></button><button class="btn btn-primary mx-2"> Subir foto <em class="fa fa-upload"></em></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que sucediÃ³ hoy?"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
